packages/@react-stately/color/src/useHexColorFieldState.ts

```diff
@@ -0,0 +1,133 @@
+ */
+
+import {Color} from './Color';
+import {ColorInput, HexColorFieldProps} from '@react-types/color';
+import {NumberFieldState} from '@react-stately/numberfield';
+import {useCallback, useState} from 'react';
+import {useControlledState} from '@react-stately/utils';
+
+export interface HexColorFieldState extends NumberFieldState {
+  colorValue: Color,
+  setInputValue: (value: string) => void
+}
+
+export const defaultMinValue = '#000000';
+export const defaultMaxValue = '#FFFFFF';
+
+export function useHexColorFieldState(
+  props: HexColorFieldProps
+): HexColorFieldState {
+  const {
+    minValue = defaultMinValue,
+    maxValue = defaultMaxValue,
+    step = 1,
+    value,
+    defaultValue,
+    onChange,
+    validationState
+  } = props;
+
+  const minColor = Color.parse(minValue);
+  const maxColor = Color.parse(maxValue);
+  const minColorInt = minColor.toHexInt();
+  const maxColorInt = maxColor.toHexInt();
+
+  const clampColor = useCallback((value: ColorInput) => {
+    try {
+      const color = Color.parse(value);
+      const colorInt = color.toHexInt();
+      if (colorInt < minColorInt) { return minColor; }
+      if (colorInt > maxColorInt) { return maxColor; }
+      return color;
+    } catch (err) {
+      return undefined;
+    }
+  }, [minColorInt, maxColorInt]);
+
+  const initialValue = clampColor(value);
+  const initialDefaultValue = clampColor(defaultValue);
+  const [colorValue, setColorValue] = useControlledState<Color>(initialValue, initialDefaultValue, onChange);
+
+  const initialInputValue = (value || defaultValue) && colorValue ? colorValue.toString('hex') : '';
+  const [inputValue, setInputValue] = useState(initialInputValue);
+
+  const increment = () => {
+    setColorValue((previousValue: Color) => {
+      const colorInt = previousValue ? previousValue.toHexInt() : minColorInt;
+      const newValue = `#${Math.min(colorInt + step, maxColorInt).toString(16).padStart(6, '0')}`;
+      const newColor = new Color(newValue);
+      setInputValue(newColor.toString('hex'));
+      return newColor;
+    });
+  };
+
+  const incrementToMax = useCallback(() => {
+    if (maxValue) {
+      setColorValue(maxColor);
+      setInputValue(maxColor.toString('hex'));
+    }
+  }, [maxValue, maxColor, setColorValue, setInputValue]);
+
+  const decrement = () => {
+    setColorValue((previousValue: Color) => {
+      if (!previousValue) { return minColor; }
+      const colorInt = previousValue.toHexInt();
+      const newValue = `#${Math.max(colorInt - step, minColorInt).toString(16).padStart(6, '0')}`;
+      const newColor = new Color(newValue);
+      setInputValue(newColor.toString('hex'));
+      return newColor;
+    });
+  };
+
+  const decrementToMin = useCallback(() => {
+    if (minValue) {
+      setColorValue(minColor);
+      setInputValue(minColor.toString('hex'));
+    }
+  }, [minValue, minColor, setColorValue, setInputValue]);
+
+  const setFieldInputValue = (value: string) => {
```

Comment:
lets ignore all characters that don't match the set `[#0-9a-fA-F]`, just trim them out before calling setInputValue, I think that'll work

Reply:
Alright! Does this mean the input won't accept any other keys at all that aren't in `[#0-9a-fA-F]`?

Reply:
correct, I can't think of any other keys that should be accepted, can you?

Reply:
Related to this change, would we add other restrictions such as typing beyond six alphanumeric characters? And allowing `#` only when it's at the beginning of the input value?

Reply:
if it's pretty easy to do, then sure

Reply:
If I paste a string with an invalid character in the middle "abcgcba", does it remove the invalid characters as well?

Reply:
- [x] Accept `[#0-9a-fA-F]` keys only
- [x] If string is pasted, remove invalid characters from entire string